# Cleaned Weather Data Script (with Date)

This script was made to obtain a cleaned version of the weather data without removing the DATE or MONTH columns

## 1. Importing Libraries and Data

In [3]:
import pandas as pd
import numpy as np
import os

In [4]:
path = r'C:\Users\kyles\CareerFoundary\Machine Learning\Achievement 2\02 Data'

In [5]:
# Import unscaled weather data
df_weather = pd.read_csv(os.path.join(path, 'ClimateWins_Dataset_original.csv'))

# Check
df_weather

DATE  MONTH  BASEL_cloud_cover  BASEL_wind_speed  BASEL_humidity  \
0      19600101      1                  7               2.1            0.85   
1      19600102      1                  6               2.1            0.84   
2      19600103      1                  8               2.1            0.90   
3      19600104      1                  3               2.1            0.92   
4      19600105      1                  6               2.1            0.95   
...         ...    ...                ...               ...             ...   
22945  20221027     10                  1               2.1            0.79   
22946  20221028     10                  6               2.1            0.77   
22947  20221029     10                  4               2.1            0.76   
22948  20221030     10                  5               2.1            0.80   
22949  20221031     10                  5               2.1            0.84   

       BASEL_pressure  BASEL_global_radiation  BASEL_precipitation  \
0              1.0180                    0.32                 0.09   
1              1.0180                    0.36                 1.05   
2              1.0180                    0.18                 0.30   
3              1.0180                    0.58                 0.00   
4              1.0180                    0.65                 0.14   
...               ...                     ...                  ...   
22945          1.0248                    1.34                 0.22   
22946          1.0244                    1.34                 0.22   
22947          1.0227                    1.34                 0.22   
22948          1.0212                    1.34                 0.22   
22949          1.0193                    1.34                 0.22   

       BASEL_snow_depth  BASEL_sunshine  ...  VALENTIA_cloud_cover  \
0                     0             0.7  ...                     5   
1                     0             1.1  ...                     7   
2                     0             0.0  ...                     7   
3                     0             4.1  ...                     7   
4                     0             5.4  ...                     3   
...                 ...             ...  ...                   ...   
22945                 0             7.7  ...                     5   
22946                 0             5.4  ...                     5   
22947                 0             6.1  ...                     5   
22948                 0             5.8  ...                     5   
22949                 0             3.2  ...                     5   

       VALENTIA_humidity  VALENTIA_pressure  VALENTIA_global_radiation  \
0                   0.88             1.0003                       0.45   
1                   0.91             1.0007                       0.25   
2                   0.91             1.0096                       0.17   
3                   0.86             1.0184                       0.13   
4                   0.80             1.0328                       0.46   
...                  ...                ...                        ...   
22945               0.82             1.0142                       1.13   
22946               0.82             1.0142                       1.13   
22947               0.82             1.0142                       1.13   
22948               0.82             1.0142                       1.13   
22949               0.82             1.0142                       1.13   

       VALENTIA_precipitation  VALENTIA_snow_depth  VALENTIA_sunshine  \
0                        0.34                    0                4.7   
1                        0.84                    0                0.7   
2                        0.08                    0                0.1   
3                        0.98                    0                0.0   
4                        0.00                    0                5.7   
...                       ...                  ...   

In [6]:
# Import weather answers data
df_answers = pd.read_csv(os.path.join(path, 'Pleasant_Weather_Prediction_Answers.csv'))
df_answers

DATE  BASEL_pleasant_weather  BELGRADE_pleasant_weather  \
0      19600101                       0                          0   
1      19600102                       0                          0   
2      19600103                       0                          0   
3      19600104                       0                          0   
4      19600105                       0                          0   
...         ...                     ...                        ...   
22945  20221027                       0                          0   
22946  20221028                       0                          0   
22947  20221029                       0                          0   
22948  20221030                       0                          0   
22949  20221031                       0                          0   

       BUDAPEST_pleasant_weather  DEBILT_pleasant_weather  \
0                              0                        0   
1                              0                        0   
2                              0                        0   
3                              0                        0   
4                              0                        0   
...                          ...                      ...   
22945                          0                        0   
22946                          0                        0   
22947                          0                        0   
22948                          0                        0   
22949                          0                        0   

       DUSSELDORF_pleasant_weather  HEATHROW_pleasant_weather  \
0                                0                          0   
1                                0                          0   
2                                0                          0   
3                                0                          0   
4                                0                          0   
...                            ...                        ...   
22945                            0                          0   
22946                            0                          0   
22947                            0                          0   
22948                            0                          0   
22949                            0                          0   

       KASSEL_pleasant_weather  LJUBLJANA_pleasant_weather  \
0                            0                           0   
1                            0                           0   
2                            0                           0   
3                            0                           0   
4                            0                           0   
...                        ...                         ...   
22945                        0                           0   
22946                        0                           0   
22947                        0                           0   
22948                        0                           0   
22949                        0                           0   

       MAASTRICHT_pleasant_weather  MADRID_pleasant_weather  \
0                                0                        0   
1                                0                        0   
2                                0                        0   
3                                0                        0   
4                                0                        0   
...                            ...                      ...   
22945                            0                        0   
22946                            0                        0   
22947                            0                        0   
22948                            0                        0   
22949                            0                        0   

       MUNCHENB_pleasant_weather  OSLO_pleasant_weather  \
0                              0                      0   
1                              0    

## 2. Data Cleaning

The following code ensures the weather data is perfectly suited for future machine learning models

- Drop Unshared Weather Stations
- Drop Observations with Missing Data
- Ensure all Observations Present for All Stations
- Check Data Shapes
- Export Cleaned Data

### 2.1 - Drop Unshared Weather Stations

In [8]:
# Create lists for weather stations for both datasets using list comprehensions
stations_in_weather = [col.split('_')[0] for col in df_weather.columns if col != 'MONTH']
stations_in_answers = [col.split('_')[0] for col in df_answers.columns]

# Find stations in weather list but not in pleasant list
missing_stations = set(stations_in_weather) - set(stations_in_answers)

# Print missing stations
print(f'Stations in weather that are not in answers: {missing_stations}')

Stations in weather that are not in answers: {'ROMA', 'TOURS', 'GDANSK'}


In [9]:
# List all columns for the above weather stations
columns_to_remove = [col for col in df_weather.columns if col.split('_')[0] in missing_stations] # This has been checked

In [10]:
# Remove these columns
df_weather.drop(columns = columns_to_remove, inplace = True)
df_weather

DATE  MONTH  BASEL_cloud_cover  BASEL_wind_speed  BASEL_humidity  \
0      19600101      1                  7               2.1            0.85   
1      19600102      1                  6               2.1            0.84   
2      19600103      1                  8               2.1            0.90   
3      19600104      1                  3               2.1            0.92   
4      19600105      1                  6               2.1            0.95   
...         ...    ...                ...               ...             ...   
22945  20221027     10                  1               2.1            0.79   
22946  20221028     10                  6               2.1            0.77   
22947  20221029     10                  4               2.1            0.76   
22948  20221030     10                  5               2.1            0.80   
22949  20221031     10                  5               2.1            0.84   

       BASEL_pressure  BASEL_global_radiation  BASEL_precipitation  \
0              1.0180                    0.32                 0.09   
1              1.0180                    0.36                 1.05   
2              1.0180                    0.18                 0.30   
3              1.0180                    0.58                 0.00   
4              1.0180                    0.65                 0.14   
...               ...                     ...                  ...   
22945          1.0248                    1.34                 0.22   
22946          1.0244                    1.34                 0.22   
22947          1.0227                    1.34                 0.22   
22948          1.0212                    1.34                 0.22   
22949          1.0193                    1.34                 0.22   

       BASEL_snow_depth  BASEL_sunshine  ...  VALENTIA_cloud_cover  \
0                     0             0.7  ...                     5   
1                     0             1.1  ...                     7   
2                     0             0.0  ...                     7   
3                     0             4.1  ...                     7   
4                     0             5.4  ...                     3   
...                 ...             ...  ...                   ...   
22945                 0             7.7  ...                     5   
22946                 0             5.4  ...                     5   
22947                 0             6.1  ...                     5   
22948                 0             5.8  ...                     5   
22949                 0             3.2  ...                     5   

       VALENTIA_humidity  VALENTIA_pressure  VALENTIA_global_radiation  \
0                   0.88             1.0003                       0.45   
1                   0.91             1.0007                       0.25   
2                   0.91             1.0096                       0.17   
3                   0.86             1.0184                       0.13   
4                   0.80             1.0328                       0.46   
...                  ...                ...                        ...   
22945               0.82             1.0142                       1.13   
22946               0.82             1.0142                       1.13   
22947               0.82             1.0142                       1.13   
22948               0.82             1.0142                       1.13   
22949               0.82             1.0142                       1.13   

       VALENTIA_precipitation  VALENTIA_snow_depth  VALENTIA_sunshine  \
0                        0.34                    0                4.7   
1                        0.84                    0                0.7   
2                        0.08                    0                0.1   
3                        0.98                    0                0.0   
4                        0.00                    0                5.7   
...                       ...                  ...   

### 2.2 - Drop Observations with Missing Data

It was determined in Exercise 2.2 that there is no missing data / NaN values.

However, there are observations which are not present for all stations (**wind_speed** and **snow_depth**).

In [14]:
# List columns with these observations
columns_to_drop = [col for col in df_weather if col.endswith('snow_depth') or col.endswith('wind_speed')]
print(columns_to_drop) # Ensure correct columns have been assigned
print(len(columns_to_drop)) # Ensure 15 columns are included

['BASEL_wind_speed', 'BASEL_snow_depth', 'DEBILT_wind_speed', 'DUSSELDORF_wind_speed', 'DUSSELDORF_snow_depth', 'HEATHROW_snow_depth', 'KASSEL_wind_speed', 'LJUBLJANA_wind_speed', 'MAASTRICHT_wind_speed', 'MADRID_wind_speed', 'MUNCHENB_snow_depth', 'OSLO_wind_speed', 'OSLO_snow_depth', 'SONNBLICK_wind_speed', 'VALENTIA_snow_depth']
15


In [15]:
# Drop columns
df_weather.drop(columns = columns_to_drop, inplace = True)
df_weather.shape # Should be down to 132 columns

(22950, 134)

### 2.3 - Ensure all Observations Present for All Stations

**pressure**, **humidity**, and **cloud_cover** are each missing from only 1 out of 15 weather stations. This data should be supplemented (copied) from a nearby weather station. Start with finding which stations are missing each observation.

In [20]:
# Create a list containing the above weather observations
missing_observations = ['cloud_cover', 'humidity', 'pressure']

In [24]:
# Make a list of all stations 
all_stations = sorted(list(set([col.split('_')[0] for col in df_weather.columns if col not in ['DATE', 'MONTH']]))) 
# Check
print(len(all_stations))

15


In [25]:
# Create dictionary for storing station and missing observation
missing_stations_by_observation = {}

for ob in missing_observations:
    # Start with selecting stations where the observation is present
    stations_covered = [col.split('_')[0] for col in df_weather.columns if col.endswith(ob)]

    # Identify missing station by subtracting covered stations from all stations 
    missing_station = set(all_stations) - set(stations_covered)

    # Store station and missing observation in dictionary
    missing_stations_by_observation[ob] = missing_station

# Print stations along with their missing observation
for ob, missing_station in missing_stations_by_observation.items():
    print(f'Station missing {ob}: {", ".join(missing_station)}')

Station missing cloud_cover: KASSEL
Station missing humidity: STOCKHOLM
Station missing pressure: MUNCHENB


In [26]:
# Check head of df_weather to assist in determining placement of new columns
df_weather.head()

DATE  MONTH  BASEL_cloud_cover  BASEL_humidity  BASEL_pressure  \
0  19600101      1                  7            0.85           1.018   
1  19600102      1                  6            0.84           1.018   
2  19600103      1                  8            0.90           1.018   
3  19600104      1                  3            0.92           1.018   
4  19600105      1                  6            0.95           1.018   

   BASEL_global_radiation  BASEL_precipitation  BASEL_sunshine  \
0                    0.32                 0.09             0.7   
1                    0.36                 1.05             1.1   
2                    0.18                 0.30             0.0   
3                    0.58                 0.00             4.1   
4                    0.65                 0.14             5.4   

   BASEL_temp_mean  BASEL_temp_min  ...  STOCKHOLM_temp_max  \
0              6.5             0.8  ...                 4.9   
1              6.1             3.3  ...                 5.0   
2              8.5             5.1  ...                 4.1   
3              6.3             3.8  ...                 2.3   
4              3.0            -0.7  ...                 4.3   

   VALENTIA_cloud_cover  VALENTIA_humidity  VALENTIA_pressure  \
0                     5               0.88             1.0003   
1                     7               0.91             1.0007   
2                     7               0.91             1.0096   
3                     7               0.86             1.0184   
4                     3               0.80             1.0328   

   VALENTIA_global_radiation  VALENTIA_precipitation  VALENTIA_sunshine  \
0                       0.45                    0.34                4.7   
1                       0.25                    0.84                0.7   
2                       0.17                    0.08                0.1   
3                       0.13                    0.98                0.0   
4                       0.46                    0.00                5.7   

   VALENTIA_temp_mean  VALENTIA_temp_min  VALENTIA_temp_max  
0                 8.5                6.0               10.9  
1                 8.9                5.6               12.1  
2                10.5                8.1               12.9  
3                 7.4                7.3               10.6  
4                 5.7                3.0                8.4  

[5 rows x 134 columns]

In [19]:
all_stations

['BASEL',
 'BELGRADE',
 'BUDAPEST',
 'DEBILT',
 'DUSSELDORF',
 'HEATHROW',
 'KASSEL',
 'LJUBLJANA',
 'MAASTRICHT',
 'MADRID',
 'MUNCHENB',
 'OSLO',
 'SONNBLICK',
 'STOCKHOLM',
 'VALENTIA']

In [27]:
# Check position of HEATHROW_temp_max to find where we need to place new KASSEL_cloud_cover column
print(f'KASSEL_cloud_cover column position: {df_weather.columns.get_loc("HEATHROW_temp_max") + 1}') # +1 to position column to the right

# Check position of MUNCHENB_humidity to find where we need to place new MUNCHENB_pressure column
print(f'MUNCHENB_pressure column position: {df_weather.columns.get_loc("MUNCHENB_humidity") + 2}') # +2 (consider Kassel's new column)

# Check position of STOCKHOLM_cloud_cover to find where we need to place new STOCKHOLM_humidity column
print(f'STOCKHOLM_humidity column position: {df_weather.columns.get_loc("STOCKHOLM_cloud_cover") + 3}') # +3 (as above)

KASSEL_cloud_cover column position: 56
MUNCHENB_pressure column position: 94
STOCKHOLM_humidity column position: 120


In [28]:
# Finally, insert new columns into df_weather
# Copy data from nearby weather stations: 
# Kassel with Ljubljana, Munchenb with Sonnblick, Stockholm with Oslo

df_weather.insert(56, 'KASSEL_cloud_cover', df_weather['LJUBLJANA_cloud_cover'])
df_weather.insert(94, 'MUNCHENB_pressure', df_weather['SONNBLICK_pressure'])
df_weather.insert(120, 'STOCKHOLM_humidity', df_weather['OSLO_humidity'])

In [29]:
# Check
df_weather.columns.tolist()

['DATE',
 'MONTH',
 'BASEL_cloud_cover',
 'BASEL_humidity',
 'BASEL_pressure',
 'BASEL_global_radiation',
 'BASEL_precipitation',
 'BASEL_sunshine',
 'BASEL_temp_mean',
 'BASEL_temp_min',
 'BASEL_temp_max',
 'BELGRADE_cloud_cover',
 'BELGRADE_humidity',
 'BELGRADE_pressure',
 'BELGRADE_global_radiation',
 'BELGRADE_precipitation',
 'BELGRADE_sunshine',
 'BELGRADE_temp_mean',
 'BELGRADE_temp_min',
 'BELGRADE_temp_max',
 'BUDAPEST_cloud_cover',
 'BUDAPEST_humidity',
 'BUDAPEST_pressure',
 'BUDAPEST_global_radiation',
 'BUDAPEST_precipitation',
 'BUDAPEST_sunshine',
 'BUDAPEST_temp_mean',
 'BUDAPEST_temp_min',
 'BUDAPEST_temp_max',
 'DEBILT_cloud_cover',
 'DEBILT_humidity',
 'DEBILT_pressure',
 'DEBILT_global_radiation',
 'DEBILT_precipitation',
 'DEBILT_sunshine',
 'DEBILT_temp_mean',
 'DEBILT_temp_min',
 'DEBILT_temp_max',
 'DUSSELDORF_cloud_cover',
 'DUSSELDORF_humidity',
 'DUSSELDORF_pressure',
 'DUSSELDORF_global_radiation',
 'DUSSELDORF_precipitation',
 'DUSSELDORF_sunshine',
 'DUSS

### 2.4 - Check Data Shapes

In [30]:
df_weather.shape

(22950, 137)

In [32]:
df_answers.shape

(22950, 16)

Both data sets have the correct shape

### 2.5 - Export Cleaned Data

In [33]:
df_weather.to_csv(os.path.join(path, 'Cleaned_Weather_Data_with_dates.csv'))

- What happens to loss and accuracy for different combinations of hyperparameters?
- Does the model converge or does the loss grow exponentially? (list the activation type used)
- How accurate is the model at recognising the stations?


### 8.1 - Changing Activation Types

**Attempt 1:**

epoch = 30, 

batch size = 16, 

number of hidden layers = 4, 

activation type = ***softmax***

**Results:** loss increased exponentially. Accuracy hovered around 16% to 11%. "Recognised" 9/15 stations poorly

**Attempt 2:**

epoch = 30, 

batch size = 16, 

number of hidden layers = 4, 

activation type = ***sigmoid***

**Results:** loss increased exponentially. Accuracy hovered around 11% to 13%. "Recognised" 1 station only (Basel)

**Attempt 3:**

epoch = 30, 

batch size = 16, 

number of hidden layers = 4, 

activation type = ***tanh***

**Results:** loss changed only from epoch 1 to epoch 4 (decreased slightly, stabalised at 22.9). Accuracy increased but stayed at 25.5% after epoch 4. "Recognised" 5 stations only.

**Attempt 4:**

epoch = 30, 

batch size = 16, 

number of hidden layers = 4, 

activation type = ***relu***

**Results:** loss became "nan" at epoch 2. Accuracy increased to 64.4% at epoch 3. This appears to be some kind of error

***Thoughts:*** perhaps the number of layers is too low right now for the model to create enough meaningful connections within the data and thus the activation types are just creating different (perhaps non-meaningful) errors because of that.

### 8.2 - Increasing Hidden Layers with Sigmoid

**Attempt 5:**

epoch = 30, 

batch size = 16, 

number of hidden layers = ***8***, 

activation type = sigmoid

**Results:** loss increased exponentially. Accuracy ≈ 13% to 14%. "Recognised" 2 stations only.

**Attempt 6:**

epoch = 30, 

batch size = 16, 

number of hidden layers = ***16***, 

activation type = sigmoid

**Results:** loss increased exponentially. Accuracy ≈ 11% to 12%. "Recognised" 1 station only.

**Attempt 7:**

epoch = 30, 

batch size = 16, 

number of hidden layers = ***32***, 

activation type = sigmoid

**Results:** loss increased exponentially. Accuracy ≈ 13% to 14%. "Recognised" 1 stations only.

***Thoughts:*** Attempts to further increase higgen layers led to similar results. Changing back to softmax at a higher starting number of hidden layers

### 8.3 - Increasing Hidden Layers with Softmax

***Summary:*** Basically, the loss function continued to grow exponentially despite the number of hidden layers. Although, 15 stations recogised after having 8 layers. Will now try tanh.

### 8.4 - Increasing Hidden Layers with Tanh

**Attempt 12:**

epoch = 30, 

batch size = 16, 

number of hidden layers = ***16***, 

activation type = tanh

**Results:** loss changed only from epoch 1 to epoch 14 (decreased slightly, stabalised at 22.8). Accuracy increased but stayed at 24.8% after epoch 14. "Recognised" 6 stations only.

***Summary of next attemtpts:*** Increasing the number of hidden layers made the model less accurate, only ever "recognising" 8 stations out of 15. loss never increased exponentially but stopped changing early into the epochs

### 8.5 - Increasing Hidden Layers with Relu

***Summary of next attempts:*** Increasing hidden layers also didn't help. Accuracy stays at 64.4%, loss goes to "nan", and matrix only recognises 1 station.

### 8.6 - Final Attempts with Softmax

Softmax seems to show the most promise for the activation type as it allows the model to seemingly identify the highest number of stations. I will experiment with the other hyperparameters using softmax before trialling out the RNN Model

***8 layers:*** 15 stations, Accuracy ≈ 10% to 14%, loss --> Exponential

***16 layers:*** 15 stations, Accuracy ≈ 10% to 14%, loss --> Exponential

***32 layers:*** 10 stations, Accuracy ≈ 10% to 14%, loss --> Exponential

***64 layers:*** 13 stations, Accuracy ≈ 12% to 14%, loss --> Exponential

***128 layers:*** 15 stations, Accuracy ≈ 10% to 13%, loss --> Exponential

Final note: After rerunning the same hyperparameters again for the purposes of obtaining screenshots, I have come to realise the model is somewhat random and may present slightly different results each time. These are usually minor (like 1% change to final accuracy, but can also be more significant, like failing to recognise multiple stations that were previously recognised. Perhaps CNN isn't so good for this